File: pyembc/host/pubsub_port.py
# Copyright 2020 Jetperch LLC
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.

import numpy as np
import struct
import logging


log = logging.getLogger(__name__)


class PayloadType:
    NULL = 0
    CSTR = 1
    U32 = 2


def _to_null(x):
    return None


def _to_cstr(x):
    return x[:-1].tobytes.decode('utf-8')


def _to_u32(x):
    if len(x) != 4:
        raise ValueError('invalid length')
    return struct.unpack('<I', x)[0]


_PAYLOAD_TYPE_FN = {
    PayloadType.NULL: _to_null,
    PayloadType.CSTR: _to_cstr,
    PayloadType.U32: _to_u32,
}


def _payload_encode(x):
    if x is None:
        return PayloadType.NULL, b''
    elif isinstance(x, str):
        return PayloadType.CSTR, x.encode('utf-8') + b'\x00'
    elif isinstance(x, int):
        if 0 <= x < (1 << 32):
            return PayloadType.U32, struct.pack('<I', x)
    raise ValueError('Unsupported payload')


class PubSubPort:

    def __init__(self, send, listener):
        """
        Implement the port compatible with the transport layer.

        :param send: The callable(port_data, msg) function to send
            data to the lower transport layer.
        :param listener: The callable(status, topic, value, port_data)
            to provide data to the higher layer.

        Provide :meth:`on_event` and :meth:`on_recv` to the
        Transport.register_port().
        """
        self._send = send
        self.listener = listener  # callable(status, topic, value, port_data)

    def on_event(self, event):
        pass

    def send(self, topic, value):
        topic_bytes = topic.encode('utf-8') + b'\x00'
        topic_len = len(topic_bytes)
        if topic_len > 32:
            raise ValueError(f'topic too long: {topic_len}')
        payload_type, payload = _payload_encode(value)
        sz = 1 + 1 + topic_len + 1 + len(payload)
        if sz > 256:
            raise ValueError(f'message too long: {sz}')
        port_data = (payload_type & 0x0f) << 8
        msg = np.empty(sz, dtype=np.uint8)
        msg[0] = (topic_len - 1)
        msg[1:topic_len + 1] = memoryview(topic_bytes)
        msg[1 + topic_len] = len(payload)
        if len(payload):
            msg[2 + topic_len:] = memoryview(payload)
        self._send(port_data, msg)

    def on_recv(self, port_data, msg):
        topic_len = (msg[0] & 0x1f) + 1
        payload_type = (port_data >> 8) & 0x0f
        topic = msg[1:topic_len].tobytes().decode('utf-8')
        payload_len = msg[1 + topic_len]
        payload = msg[2 + topic_len:]
        if len(payload) != payload_len:
            log.warning('Invalid payload')
            return
        payload_fn = _PAYLOAD_TYPE_FN[payload_type]
        x = payload_fn(payload)
        log.info("recv(topic='%s', value=%s, port_data=%s)", topic, x, port_data)
        if callable(self.listener):
            self.listener(topic, x)

File: src/stream/pubsub_port.c
/*
 * Copyright 2021 Jetperch LLC
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

// #define EMBC_LOG_LEVEL EMBC_LOG_LEVEL_INFO
#include "embc/stream/pubsub_port.h"
#include "embc/bbuf.h"
#include "embc/ec.h"
#include "embc/log.h"
#include "embc/cstr.h"

enum state_e {
    ST_INIT,
    ST_ACTIVE,
};

struct embc_pubsubp_s {
    uint8_t state;
    uint8_t port_id;
    uint8_t mode;
    struct embc_pubsub_s * pubsub;
    struct embc_transport_s * transport;
    uint8_t msg[EMBC_FRAMER_PAYLOAD_MAX_SIZE];
};

const char EMBC_PUBSUBP_META[] = "{\"type\":\"pubsub\"}";

struct embc_pubsubp_s * embc_pubsubp_initialize(struct embc_pubsub_s * pubsub, enum embc_pubsubp_mode_e mode) {
    struct embc_pubsubp_s * self = embc_alloc_clr(sizeof(struct embc_pubsubp_s));
    EMBC_ASSERT_ALLOC(self);
    self->mode = mode;
    self->pubsub = pubsub;
    return self;
}

void embc_pubsubp_finalize(struct embc_pubsubp_s * self) {
    if (self) {
        embc_free(self);
    }
}

static int32_t connect(struct embc_pubsubp_s * self) {
    int32_t rc = 0;
    if (self->state != ST_INIT) {
        return 0;
    }

    if (self->mode == EMBC_PUBSUBP_MODE_UPSTREAM) {
        rc = embc_pubsub_subscribe_link(self->pubsub, "",
                                        (embc_pubsub_subscribe_fn) embc_pubsubp_on_update,
                                        self);
        if (rc) {
            EMBC_LOGE("pubsub subscribe failed");
            return rc;
        }
    }
    self->state = ST_ACTIVE;

    return 0;
}

int32_t embc_pubsubp_transport_register(struct embc_pubsubp_s * self,
                                        uint8_t port_id,
                                        struct embc_transport_s * transport) {
    self->port_id = port_id;
    self->transport = transport;
    int32_t rc = embc_transport_port_register(
            transport,
            port_id,
            EMBC_PUBSUBP_META,
            (embc_transport_event_fn) embc_pubsubp_on_event,
            (embc_transport_recv_fn) embc_pubsubp_on_recv,
            self);
    if (rc) {
        return rc;
    }
    return connect(self);
}

void embc_pubsubp_on_event(struct embc_pubsubp_s *self, enum embc_dl_event_e event) {
    (void) self;
    (void) event;
}

void embc_pubsubp_on_recv(struct embc_pubsubp_s *self,
                          uint8_t port_id,
                          enum embc_transport_seq_e seq,
                          uint16_t port_data,
                          uint8_t *msg, uint32_t msg_size) {
    if (!self->pubsub) {
        return;
    }
    if (port_id != self->port_id) {
        EMBC_LOGW("port_id mismatch: %d != %d", (int) port_id, (int) self->port_id);
        return;
    }
    if (seq != EMBC_TRANSPORT_SEQ_SINGLE) {
        EMBC_LOGW("invalid seq: %d", (int) seq);
        return;
    }
    uint8_t type = (port_data >> 8) & 0xff;
    uint8_t dtype = type & EMBC_PUBSUB_DTYPE_MASK;
    if ((port_data & 0x00ff) != 0) {
        EMBC_LOGW("invalid port_data: %d", (int) port_data);
        return;
    }
    if (msg_size < 3) {
        EMBC_LOGW("msg too small");
        return;
    }
    uint8_t topic_len = (msg[0] & 0x1f) + 1;  // 32 max
    char * topic = (char *) (msg + 1);

    uint32_t sz = 2 + topic_len;
    if (msg_size < sz) {
        EMBC_LOGW("msg too small: %d < %d", (int) msg_size, (int) sz);
        return;
    }
    if (topic[topic_len - 1]) {
        EMBC_LOGW("topic invalid");
        return;
    }

    uint8_t *payload = msg + sz - 1;
    uint8_t payload_len = *payload++;
    sz += payload_len;
    if (!payload_len || (msg_size < sz)) {
        EMBC_LOGW("msg too small: %d < %d", (int) msg_size, (int) sz);
        return;
    }

    // parse message
    struct embc_pubsub_value_s value;
    value.type = type;
    switch (dtype) {
        case EMBC_PUBSUB_DTYPE_NULL:
            break;
        case EMBC_PUBSUB_DTYPE_STR:
            if (payload[payload_len - 1]) {
                EMBC_LOGW("invalid payload string");
            } else {
                EMBC_LOGW("CSTR not fully supported yet");
                // WARNING todo support this correctly
                value.value.str = (char *) payload;  // not const and persistent!!!
            }
            break;
        case EMBC_PUBSUB_DTYPE_U32:
            if (payload_len != 4) {
                EMBC_LOGW("invalid payload u32");
            } else {
                value.value.u32 = EMBC_BBUF_DECODE_U32_LE(payload);
            }
            break;

        default:
            EMBC_LOGW("unsupported type: %d", (int) dtype);
            return;
    }
    embc_pubsub_publish(self->pubsub, topic, &value,
                        (embc_pubsub_subscribe_fn) embc_pubsubp_on_update, self);
}

uint8_t embc_pubsubp_on_update(struct embc_pubsubp_s *self,
                               const char * topic, const struct embc_pubsub_value_s * value) {
    if (!self->transport) {
        return 0;
    }
    uint8_t topic_len = 0;
    char * p = (char *) (self->msg + 1);
    while (*topic) {
        if (topic_len >= (EMBC_PUBSUB_TOPIC_LENGTH_MAX - 1)) {
            EMBC_LOGW("topic too long");
            return EMBC_ERROR_PARAMETER_INVALID;
        }
        *p++ = *topic++;
        ++topic_len;
    }
    *p++ = 0;       // add string terminator
    ++topic_len;

    uint8_t dtype = value->type & EMBC_PUBSUB_DTYPE_MASK;
    uint8_t dflag = value->type & EMBC_PUBSUB_DFLAG_MASK;
    dflag &= ~EMBC_PUBSUB_DFLAG_CONST;  // remove const, only applicable locally
    uint16_t port_data = 0 | (((uint16_t) (dflag | dtype)) << 8);

    self->msg[0] = ((topic_len - 1) & 0x1f);
    uint8_t payload_sz_max = (uint8_t) (sizeof(self->msg) - ((uint8_t *) (p + 1) - self->msg));
    uint8_t payload_sz = 0;
    uint8_t * hdr = (uint8_t *) p++;

    switch (dtype) {
        case EMBC_PUBSUB_DTYPE_NULL:
            break;
        case EMBC_PUBSUB_DTYPE_STR:  // intentional fall-through
        case EMBC_PUBSUB_DTYPE_JSON: {
            const char * s = value->value.str;
            while (*s) {
                if (payload_sz >= (payload_sz_max - 1)) {
                    EMBC_LOGW("payload full");
                    return EMBC_ERROR_PARAMETER_INVALID;
                }
                *p++ = *s++;
                ++payload_sz;
            }
            *p++ = 0;       // add string terminator
            ++payload_sz;
            break;
        }
        case EMBC_PUBSUB_DTYPE_BIN: {
            if (payload_sz_max < value->size) {
                EMBC_LOGW("payload full");
                return EMBC_ERROR_PARAMETER_INVALID;
            }
            const uint8_t * s = value->value.bin;
            for (uint8_t sz = 0; sz < value->size; ++sz) {
                *p++ = *s++;
                ++payload_sz;
            }
            break;
        }
        case EMBC_PUBSUB_DTYPE_U32:
            if (payload_sz_max < 4) {
                EMBC_LOGW("payload full");
                return EMBC_ERROR_PARAMETER_INVALID;
            } else {
                EMBC_BBUF_ENCODE_U32_LE(p, value->value.u32);
                payload_sz = 4;
            }
            break;
        default:
            EMBC_LOGW("unsupported type: %d", (int) value->type);
            return EMBC_ERROR_PARAMETER_INVALID;
    }
    *hdr = (uint8_t) payload_sz;
    embc_transport_send(self->transport, self->port_id, EMBC_TRANSPORT_SEQ_SINGLE,
                        port_data, self->msg, 2 + topic_len + payload_sz);
    return 0;
}
